### PR TITLE
Make `cargo fix` also remove unneeded braces when removing all but one nested import

### PR DESCRIPTION
**Pull request by ElectrifyPro**
In the current / nightly version of `rustc`, `cargo fix` will not remove braces in nested imports, if all but one nested import is removed. For example, running `cargo fix` (rustc 1.71) on:

```rs
use std::sync::{mpsc, Arc};
use std::thread::{self};
use std::time::Duration;
use std::collections::{
    hash_map::{Drain, HashMap, IntoIter},
    HashSet,
};

use std::cell::{RefCell, RefMut};

fn main() {
    let _x = Arc::new(5);
    let _y: HashMap<(), ()> = HashMap::new();
    let _z = RefCell::new(());
}
```

Will result in:

```rs
use std::sync::{Arc};


use std::collections::{
    hash_map::{HashMap},
};

use std::cell::{RefCell};

fn main() {
    let _x = Arc::new(5);
    let _y: HashMap<(), ()> = HashMap::new();
    let _z = RefCell::new(());
}
```

Which is not ideal. This PR modifies the output of `rustc` in these scenarios to also include the braces of the nested imports where applicable. The result of `cargo fix` with these changes is:

```rs
use std::sync::Arc;


use std::collections::{
    hash_map::HashMap,
};

use std::cell::RefCell;

fn main() {
    let _x = Arc::new(5);
    let _y: HashMap<(), ()> = HashMap::new();
    let _z = RefCell::new(());
}
```

~~I believe I need to write tests for this, but I'm not exactly sure where to put them~~

### CODE REVIEW

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2537,11 +2537,20 @@ pub enum UseTreeKind {
     /// `use prefix` or `use prefix as rename`
     Simple(Option<Ident>),
     /// `use prefix::{...}`
-    Nested(ThinVec<(UseTree, NodeId)>),
+    Nested(UseTreeNested),
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/ast.rs:2540]: Why not just add a `Span` member here? Then you don't need to do all those `nested` -> `nested.items` changes below, and you can just ignore the span where it's not used with `_`.

Comment (ElectrifyPro) [reply compiler/rustc_ast/src/ast.rs:2540]: Yes, I think that would be much better.

File: compiler/rustc_resolve/src/check_unused.rs [hunk 2]
```diff
@@ -280,6 +289,32 @@ fn calc_unused_spans(
...
             } else {
+                // If removing the nested imports leaves one import remaining, expand the removal
+                // span to include the braces, so that rustfix can remove them
+                if nested.len() - num_imports_removed == 1 {
+                    if last_nested_import_removed {
+                        // Expand the last import's span to include the braces, then add the first
+                        // brace span
+                        let new_span =
+                            to_remove.last().unwrap().shrink_to_lo().to(span.shrink_to_hi());
+                        *to_remove.last_mut().unwrap() = new_span;
```
Comment (WaffleLapkin) [inline compiler/rustc_resolve/src/check_unused.rs:300]:
```suggestion
                        let last = to_remove.last_mut().unwrap();
                        *last = last.shrink_to_lo().to(span.shrink_to_hi());
```

Relatively, can you explain why `unwrap` here is fine? Is it because the condition above asserts that there was something unused which implies `to_remove` is not empty? The code is a bit hard to follow...

Also, why is `shrink_to_lo` necessary? Can to remove span ever overlap with the end of `}`?

Comment (WaffleLapkin) [reply compiler/rustc_resolve/src/check_unused.rs:300]: Hm, would it maybe be easier to push the `}` span here, instead of modifying the last removed span? Or are there significant downsides to that?

Comment (ElectrifyPro) [reply compiler/rustc_resolve/src/check_unused.rs:300]: > Relatively, can you explain why `unwrap` here is fine? Is it because the condition above asserts that there was something unused which implies `to_remove` is not empty? The code is a bit hard to follow...

Yes, it is because that the number of nested imports in the block `nested.len()` is greater than the number of removed imports `num_imports_removed`. I may need to refactor it if the intent is not clear.

> Also, why is `shrink_to_lo` necessary? Can to remove span ever overlap with the end of `}`?

Admittedly, I did not completely understand the difference between `span.shrink_to_lo().to(hi)` instead of `span.to(hi)`. Sorry about that. What is the semantic difference? For example, if I have a span that spans bytes 10..15, then I call `span.to(hi_span)` and `hi_span` spans 20..21, will the result be the same (10..21) if I had `span.shrink_to_lo().to(hi_span)`?

> Hm, would it maybe be easier to push the `}` span here, instead of modifying the last removed span? Or are there significant downsides to that?

Yes, I think it would be easier overall to just push the spans of the braces, without modifying the exisiting removed spans. I only wrote it like this because I observed the code above mine trying to merge adjacent spans, so i thought it was significant.

Comment (WaffleLapkin) [reply compiler/rustc_resolve/src/check_unused.rs:300]: > I may need to refactor it if the intent is not clear.

If you have ideas how to improve this code, then go ahead! Although I would suggest to do it in a follow-up PR, to keep the diffs easier to review :)

----

Regarding `Span::to`... Some `Span` combinators work weirdly in case of overlapping spans, although I can't remember the exact details. For `to` it actually doesn't matter as it `min`/`max`es:
https://github.com/rust-lang/rust/blob/b7d8c88b64843d05761576aa093a34a8bc176817/compiler/rustc_span/src/lib.rs#L844-L845

So you can safely remove both shrinks, actually.

----

> Yes, I think it would be easier overall to just push the spans of the braces, without modifying the existing removed spans. But it does have some minor advantages. If someone writes [...] The extra whitespace from start of nested import to the first import, is considered and will be removed: [...]

I don't think it matters, while it's nice to keep formatting, I think it's ok to depend on rustfmt to fixup our suboptimal output (also I'd guess that whitespace around `{}` in use trees is rare). So I'd prefer the simpler code, although I don't have strong feelings here.

Comment (ElectrifyPro) [reply compiler/rustc_resolve/src/check_unused.rs]: I think the new code should be cleaner and easier to understand. I was able to remove one pair of span shrinks but in my testing, I couldn't remove the rest without the test failing.

**Comment by cjgillot** [comment]
This is computes spans from post-expansion AST and HIR, could you add a test with macro expansions? Ideally for each `Span::to`, we'd need a test where the 2 operands are not from the same expansion. For instance:
```
macro_rules! foo {
  ($e:path) => { use crate::{$e, bar, foobar!()}; }
}
```

The current suggestion is already correct, and not horrifyingly ugly. This PR adds quite many spans for cosmetics only. Isn't rustfmt the better tool to cleanup after automated suggestions?

**Comment by ElectrifyPro** [comment]
> The current suggestion is already correct, and not horrifyingly ugly. This PR adds quite many spans for cosmetics only. Isn't rustfmt the better tool to cleanup after automated suggestions?

Yes, I believe this could be a better issue to solve in `rustfmt`. Though, I did discussed this further [here](https://github.com/rust-lang/rust/pull/111045#discussion_r1181760209). To summarize, I could not use `rustfmt` to fix my (annoyingly) specific import pattern, and because I noticed that `rustc` already emits some simple formatting suggestions (like unneeded parentheses around expressions), I wanted to add a fix that would allow me to fix my unused imports without breaking my format.

> This is computes spans from post-expansion AST and HIR, could you add a test with macro expansions? Ideally for each `Span::to`, we'd need a test where the 2 operands are not from the same expansion.

👍 Though before I commit this, I would like to get your opinion on my use case.

**Comment by flip1995** [comment]
`rustfmt` already removes braces around single imports. You can try it on the [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=877f86df90004f03c906bdab9a8f4a03). `rustfmt` turns your example into

```rust
use std::sync::Arc;

use std::collections::hash_map::HashMap;

use std::cell::RefCell;
```

Which is a bit better to what the new suggestion produces. At least in Clippy we assume, that `rustfmt` is run after applying Clippy lints (with `--fix`). So we don't care about perfectly formatted suggestions. I think we can also make this assumption about `rustc`.

**Comment by ElectrifyPro** [comment]
> EDIT: I just saw your comment https://github.com/rust-lang/rust/pull/111045#discussion_r1181760209 and also want to address this: If people should not use rustfmt or hand-format some code, then formatting is on them and not on the tool providing the suggestions IMHO.

Yes, `rustfmt` cannot possibly please everyone 😁

Then I shall close this PR unless if there is any more useful discussion.
